source/docs/h2_metadata.md

```diff
@@ -11,12 +11,22 @@ are represented by a map.
 Note: the metadata implementation is still in progress, and the doc is in draft
 version.
 
-### Limitation
+### Limitation and conditions.
 
 For ease of implementation and compatibility purposes, metadata will only be
 supported in HTTP/2. Metadata sent in any other protocol should result in protocol
 errors or be ignored.
 
+To simplify the implementation, we don't allow metadata frames to carry end of
+stream flag. Because metadata frames must be associated with an existing frame, users must
+ensure metadata frames to be received before the end of stream is received by the
+peer.
+
+Metadata associated with a response can be sent before response headers, after response headers,
+between response data or after response data. If metadata frames have to be sent last,
+users must put the end of stream in an empty data frame and send the empty data frame after metadata frames.
+TODO(soya3129): add more conditions for metadata in requests.
```

Comment:
One more condition - I believe headers and trailers have per-header/per-trailer size limits and since you can only have one of each that is sufficient.  Technically you can have infinite metadata frames and they bypass flow control so I believe in-Envoy space we either need to cap the number of metadata frames (can have high and add a config option if anyone needs to go above) or to make sure the total size of metadata is less than some metadata size bounds rather than having each individual metadata being limited but allowing infinite frames.  Totally fine to do this in a follow-up but as I believe it's a DoS OOM vector I'd like an explicit TODO for this particular one it so we don't lose track of it :-)

Reply:
A very good point! Added a stream level condition in MetadataDecoder. If needed, we can add to MetadataEncoder or at connection level in the future.